### DISCUSSION TITLE
Migration to Microbot v2

### DISCUSSION THREAD

**Discussion by PatLub85** [micropython/micropython#13517] [General]
Hi,

I have this Microbot v1 program which works and which I would like to upgrade to v2
```
from microdot_asyncio import Microdot, Response, send_file
from microdot_utemplate import render_template
app = Microdot()
Response.default_content_type = 'text/html'
@app.route('/')
async def index(request):
    return render_template('index2.html', led_value=led_module.get_value())
```
Here is what I did following the informations from https://microdot.readthedocs.io/en/latest/migrating.html
```
from microdot import Microdot,Response,send_file
from template import Template
app = Microdot()
Response.default_content_type = 'text/html'
@app.route('/')
async def index(request):
    return Template('index2.html', led_value=led_module.get_value()).render() # line 36
```
I have the error message :
```
Traceback (most recent call last):
  File "microdot.py", line 1327, in dispatch_request
  File "microdot.py", line 41, in invoke_handler
  File "<stdin>", line 36, in index
TypeError: unexpected keyword argument 'led_value'
```
If I modify the place of the led_value variable
```
async def index(request):
    led_value=led_module.get_value()
    return Template('index2.html', led_value).render()
```
I have another error message :
```
Traceback (most recent call last):
  File "microdot.py", line 1327, in dispatch_request
  File "microdot.py", line 41, in invoke_handler
  File "<stdin>", line 37, in index
TypeError: function takes 2 positional arguments but 3 were given
```

Do you see what's the problem?

**Comment by sophiedegran** [comment]
Hi
Please read the migration guide from microdot to 1.x from 2.x in microdot git..
Miguel have done major updates and it work particullary well, quick with low memory
They're also a 'discussion' open on the microdot git
Hope that help
--> return await Template ('index2.html').render_async (led_value=led_value)

**Reply by PatLub85** [reply]
I have read this document several times and I have not found the solution to my problem.

**Reply by sophiedegran** [reply]
Examples also changed...
all is now in async mode
> return await Template ('index2.html').render_async (led_value=led_value)

**Reply by PatLub85** [reply]
It’s a headache!! but it works. Microdot is great, but there are few implementations on the web, and even fewer in v2
Thanks a lot
